src/datatable/datatable.jl

```diff
@@ -74,117 +74,105 @@ type DataTable <: AbstractDataTable
     colindex::Index
 
     function DataTable(columns::Vector{Any}, colindex::Index)
-        ncols = length(columns)
-        if ncols > 1
-            nrows = length(columns[1])
-            equallengths = true
-            for i in 2:ncols
-                equallengths &= length(columns[i]) == nrows
+        if length(columns) == length(colindex) == 0
+            return new(Vector{Any}(0), Index())
+        elseif length(columns) != length(colindex)
+            throw(DimensionMismatch("Number of columns ($(length(columns))) and number of column names ($(length(colindex))) are not equal"))
+        end
+        lengths = [isa(col, AbstractArray) ? length(col) : 1 for col in columns]
+        minlen, maxlen = extrema(lengths)
+        if minlen == 0 && maxlen == 0
+            return new(columns, colindex)
+        elseif minlen != maxlen || minlen == maxlen == 1
+            # recycle scalars
+            for i in 1:length(columns)
+                typeof(columns[i]) <: AbstractArray && continue
+                columns[i] = fill(columns[i], maxlen)
+                lengths[i] = maxlen
             end
-            if !equallengths
-                msg = "All columns in a DataTable must be the same length"
-                throw(ArgumentError(msg))
+            uls = unique(lengths)
+            if length(uls) != 1
+                strnames = string.(names(colindex))
+                estring = ["column length ($(uls[i])) for column(s) ($(join(strnames[find(uls .== u)], ", ")))"
+                           for (i,u) in enumerate(uls)]
+                throw(DimensionMismatch(join(estring, " is incompatible with ")))
             end
         end
-        if length(colindex) != ncols
-            msg = "Columns and column index must be the same length"
-            throw(ArgumentError(msg))
+        for (i,c) in enumerate(columns)
+            if isa(c, Range)
+                columns[i] = collect(c)
+            elseif !isa(c, AbstractVector) && isa(c, AbstractArray)
+                    throw(DimensionMismatch("columns must be 1-dimensional"))
+            else
+                columns[i] = c
+            end
         end
         new(columns, colindex)
     end
 end
 
 function DataTable(; kwargs...)
-    result = DataTable(Any[], Index())
-    for (k, v) in kwargs
-        result[k] = v
+    if length(kwargs) == 0
```

Comment:
Can you remind me why you need to handle this case explicitly? Maybe a comment would be in order?